src/data/datasets.py

```diff
@@ -0,0 +1,402 @@
+                num_words = len(ix_to_word)
+                del captions
+                if split == "train":
+                    return train_captions, ix_to_word, word_to_ix, num_words
+                return test_captions, ix_to_word, word_to_ix, num_words
+
+        else:  # if not, process the captions and save them
+            train_files = self.load_filenames(data_dir, "train")
+            test_files = self.load_filenames(data_dir, "test")
+
+            train_captions_tokenized = self.get_tokenized_captions(
+                data_dir, train_files
+            )
+            test_captions_tokenized = self.get_tokenized_captions(
+                data_dir, test_files
+            )  # we need both train and test captions to build the vocab
+
+            (
+                train_captions,
+                test_captions,
+                ix_to_word,
+                word_to_ix,
+                num_words,
+            ) = self.build_vocab(  # type: ignore
+                train_captions_tokenized, test_captions_tokenized, split
+            )
+            vocab_list = [train_captions, test_captions, ix_to_word, word_to_ix]
+            with open(captions_ckpt_path, "wb") as ckpt_file:
+                pickle.dump(vocab_list, ckpt_file)
+
+            if split == "train":
+                return train_captions, ix_to_word, word_to_ix, num_words
+            if split == "test":
+                return test_captions, ix_to_word, word_to_ix, num_words
+            raise ValueError("Invalid split. Please use 'train' or 'test'")
+
+    def build_vocab(
+        self, tokenized_captions_train: list, tokenized_captions_test: list  # type: ignore
+    ) -> Any:
+        """
+        Helper function which builds the vocab dicts.
+        :param tokenized_captions_train: list containing all the
+        train tokenized captions in the dataset. This is list of lists.
+        :param tokenized_captions_test: list containing all the
+        test tokenized captions in the dataset. This is list of lists.
+        :return train_captions_int: list of all captions in training,
+        where each word is replaced by its index in the vocab
+        :return test_captions_int: list of all captions in test,
+        where each word is replaced by its index in the vocab
+        :return ix_to_word: dictionary mapping index to word
+        :return word_to_ix: dictionary mapping word to index
+        :return num_words: number of unique words in the vocabulary
+        """
+        vocab = defaultdict(int)  # type: ignore
+        total_captions = tokenized_captions_train + tokenized_captions_test
+        for caption in total_captions:
+            for word in caption:
+                vocab[word] += 1
+
+        # sort vocab dict by frequency in descending order
+        vocab = sorted(vocab.items(), key=lambda x: x[1], reverse=True)  # type: ignore
+
+        ix_to_word = {}
+        word_to_ix = {}
+        ix_to_word[0] = "<end>"
```

Comment:
is there a start-of-sequence token?

Reply:
no, we only use this token which also acts as the padding token. Not sure whether if we require any start of sequence token? Their implementation did not use any.. so should we go with this?

Reply:
that is really strange, usually we want to have at least 4 special tokens:

1. start of sequence
2. end of sequence
3. unknown token
4. padding token

and here they only use one...

Reply:
> i don't think there will be any UNK token needed here
But what about casual inference when we deploy it using the application? Certainly there will be some words that user writes and that the model is unaware of

> it will still be "end" and it acts as padding token.
I might agree on that for now but I never saw people have same token for such purposes

> maybe mostly for seq2seq task
Agreed

Reply:
> But what about casual inference when we deploy it using the application? Certainly there will be some words that user writes and that the model is unaware of

right lol, if we are gonna deploy it as an app, then yup we might require UNK token in that case..

> I might agree on that for now but I never saw people have same token for such purposes
I think the same reason why we don't need a start token for our task, we might also do away with end token as there is no specific need to distinguish where text starts and ends right? even if you deploy as application, we can figure out if the text has ended or not.